### PR TITLE
sbt bridge reporter

### PR DESCRIPTION
**Pull request by MasseGuillaume**
replaces #1528

`dotty-bridge/scripted compilerReporter/simple`

### CODE REVIEW

File: bridge/src/main/scala/xsbt/DelegatingReporter.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        case _: Reporter.Error   => xsbti.Severity.Error
+        case _: Reporter.Warning => xsbti.Severity.Warn
+        case _                   => xsbti.Severity.Info
+      }
+    val pos  = 
+      if(d.pos.exists) Some(d.pos)
+      else None
+
+    val file = 
+      if(d.pos.source.file.exists) {
```
Comment (smarter) [inline bridge/src/main/scala/xsbt/DelegatingReporter.scala]: Style: `if (`, not `if(` (same issue in other parts of this file)

File: bridge/src/main/scala/xsbt/DelegatingReporter.scala [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  override def printSummary(implicit ctx: Context): Unit = delegate.printSummary()
+
+  def doReport(d: Diagnostic)(implicit ctx: Context): Unit = {
+    val severity = 
+      d match {
+        case _: Reporter.Error   => xsbti.Severity.Error
+        case _: Reporter.Warning => xsbti.Severity.Warn
+        case _                   => xsbti.Severity.Info
+      }
+    val pos  =
```
Comment (smarter) [inline bridge/src/main/scala/xsbt/DelegatingReporter.scala]: remove extra space between `pos` and `=`

File: bridge/src/main/scala/xsbt/DelegatingReporter.scala [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        val r = d.pos.source.file.file 
+        if(r == null) None
+        else Some(r)
+      }
+      else None
+
+    val offset0 = pos.map(_.point)
+
+    val position = new xsbti.Position {
+      def line: Maybe[Integer] = maybe(pos.map(_.line))
```
Comment (smarter) [inline bridge/src/main/scala/xsbt/DelegatingReporter.scala]: `Maybe[Int]` and `Maybe[Integer]` are the same thing so just use `Maybe[Int]`

File: bridge/src/main/scala/xsbt/DelegatingReporter.scala [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    delegate.log(position, d.message, severity)
+  }
+
+  private[this] def maybe[T](opt: Option[T]): Maybe[T] = opt match { 
+    case None => Maybe.nothing[T]
+    case Some(s) => Maybe.just[T](s)
+  }
+  import java.lang.{ Integer => I }
+  private[this] def maybeInt(opt: Option[Int]): Maybe[I] = opt match {
```
Comment (smarter) [inline bridge/src/main/scala/xsbt/DelegatingReporter.scala]: This method is unnecessary, just use `maybe` above

File: bridge/src/main/scala/xsbt/DelegatingReporter.scala [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        else Some(r)
+      }
+      else None
+
+    val offset0 = pos.map(_.point)
+
+    val position = new xsbti.Position {
+      def line: Maybe[Integer] = maybe(pos.map(_.line))
+      def lineContent(): String = pos.map(_.lineContent).getOrElse("")
+      def offset(): xsbti.Maybe[Integer] = maybeInt(offset0)
```
Comment (smarter) [inline bridge/src/main/scala/xsbt/DelegatingReporter.scala]: `xsbti.Maybe` is imported so you can just write `Maybe[...]`

File: bridge/src/main/scala/xsbt/DelegatingReporter.scala [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        if(r == null) None
+        else Some(r)
+      }
+      else None
+
+    val offset0 = pos.map(_.point)
+
+    val position = new xsbti.Position {
+      def line: Maybe[Integer] = maybe(pos.map(_.line))
+      def lineContent(): String = pos.map(_.lineContent).getOrElse("")
```
Comment (smarter) [inline bridge/src/main/scala/xsbt/DelegatingReporter.scala]: Drop the empty param list `()` (same below)

File: bridge/src/main/scala/xsbt/DelegatingReporter.scala [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        case _                   => xsbti.Severity.Info
+      }
+    val pos  = 
+      if(d.pos.exists) Some(d.pos)
+      else None
+
+    val file = 
+      if(d.pos.source.file.exists) {
+        val r = d.pos.source.file.file 
+        if(r == null) None
```
Comment (smarter) [inline bridge/src/main/scala/xsbt/DelegatingReporter.scala]: Just write `Option(d.pos.source.file.file)`

File: project/Build.scala [hunk 8]
```diff
@@ -207,11 +207,28 @@ object DottyBuild extends Build {
...
+        val classVersion = System.getProperty("java.class.version")
+        val sbtV = sbtVersion.value
+        val home = System.getProperty("user.home")
+        val org = organization.value
+        val artifact = moduleName.value
+
+        IO.delete(file(home) / ".ivy2" / "cache" / "org.scala-sbt" / s"$org-$artifact-$dottyBridgeVersion-bin_${dottyVersion}__$classVersion")
+        IO.delete(file(home) / ".sbt"  / "boot" / "scala-2.10.6" / "org.scala-sbt" / "sbt" / sbtV / s"$org-$artifact-$dottyBridgeVersion-bin_${dottyVersion}__$classVersion")
+      },
+      publishLocal <<= publishLocal.dependsOn(cleanBridge),
```
Comment (sjrd) [inline project/Build.scala]: Awesome, finally!

Comment (MasseGuillaume) [inline project/Build.scala]: oki `publishLocal := (publishLocal.dependsOn(cleanBridge)).value`

File: bridge/src/main/scala/xsbt/CompilerInterface.scala [hunk 9]
```diff
@@ -50,12 +50,12 @@ class CachedCompilerImpl(args: Array[String], output: Output, resident: Boolean)
...
-    run(sources.toList, changes, callback, log, progress)
+    run(sources.toList, changes, callback, log, delegate, progress)
   }
-  private[this] def run(sources: List[File], changes: DependencyChanges, callback: AnalysisCallback, log: Logger, compileProgress: CompileProgress): Unit = {
+  private[this] def run(sources: List[File], changes: DependencyChanges, callback: AnalysisCallback, log: Logger, delegate: Reporter, compileProgress: CompileProgress): Unit = {
     debug(log, args.mkString("Calling Dotty compiler with arguments  (CompilerInterface):\n\t", "\n\t", ""))
-    val ctx = (new ContextBase).initialCtx.fresh
-      .setSbtCallback(callback)
+    val freshContext = (new ContextBase).initialCtx.fresh
+    val ctx = freshContext.setSbtCallback(callback).setReporter(DelegatingReporter(delegate))
```
Comment (MasseGuillaume) [inline bridge/src/main/scala/xsbt/CompilerInterface.scala]: it crashes the compiler if I don't write it like this.

File: bridge/src/main/scala/xsbt/DelegatingReporter.scala (hunk 4 above)
Comment (MasseGuillaume) [inline bridge/src/main/scala/xsbt/DelegatingReporter.scala]: the compiler fails to infer with only `maybe`

File: bridge/src/main/scala/xsbt/CompilerInterface.scala (hunk 9 above)
Comment (smarter) [inline bridge/src/main/scala/xsbt/CompilerInterface.scala]: OK, could you come up with a minimal example and open an issue? and leave a comment to indicate that you're working around a crash with the issue number

File: bridge/src/main/scala/xsbt/DelegatingReporter.scala (hunk 4 above)
Comment (smarter) [inline bridge/src/main/scala/xsbt/DelegatingReporter.scala]: then write `maybe[Int]`, please open an issue for this too if this is an issue specific to dotty

File: bridge/src/main/scala/xsbt/CompilerInterface.scala (hunk 9 above)
Comment (MasseGuillaume) [inline bridge/src/main/scala/xsbt/CompilerInterface.scala]: strange, I just tested and it works fine.

File: bridge/src/main/scala/xsbt/DelegatingReporter.scala [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+import dotty.tools._
+import dotc._
+import reporting._
+import core.Contexts._
+
+import xsbti.Maybe
+
+object DelegatingReporter {
+  def apply(delegate: xsbti.Reporter) = new DelegatingReporter(delegate)
```
Comment (MasseGuillaume) [inline bridge/src/main/scala/xsbt/DelegatingReporter.scala]: done

File: bridge/src/main/scala/xsbt/DelegatingReporter.scala (hunk 2 above)
Comment (MasseGuillaume) [inline bridge/src/main/scala/xsbt/DelegatingReporter.scala]: done

File: bridge/src/main/scala/xsbt/DelegatingReporter.scala (hunk 1 above)
Comment (MasseGuillaume) [inline bridge/src/main/scala/xsbt/DelegatingReporter.scala]: done

File: bridge/src/main/scala/xsbt/DelegatingReporter.scala (hunk 7 above)
Comment (MasseGuillaume) [inline bridge/src/main/scala/xsbt/DelegatingReporter.scala]: done

File: bridge/src/main/scala/xsbt/DelegatingReporter.scala (hunk 4 above)
Comment (MasseGuillaume) [inline bridge/src/main/scala/xsbt/DelegatingReporter.scala]: scalac:

```scala
trait Position {
  def offset(): Integer  
}

object A {
  val p = new Position {
    def offset(): Int = 1
  }
}
```

overriding method offset in trait Position of type ()Integer;
 method offset has incompatible type

File: bridge/src/main/scala/xsbt/DelegatingReporter.scala (hunk 6 above)
Comment (MasseGuillaume) [inline bridge/src/main/scala/xsbt/DelegatingReporter.scala]: done

File: bridge/src/main/scala/xsbt/DelegatingReporter.scala (hunk 5 above)
Comment (MasseGuillaume) [inline bridge/src/main/scala/xsbt/DelegatingReporter.scala]: done

File: bridge/src/main/scala/xsbt/DelegatingReporter.scala (hunk 3 above)
Comment (MasseGuillaume) [inline bridge/src/main/scala/xsbt/DelegatingReporter.scala]: see https://github.com/lampepfl/dotty/pull/1530/files#r80051828

File: project/Build.scala [hunk 11]
```diff
@@ -207,11 +207,28 @@ object DottyBuild extends Build {
     ).
     settings(publishing)
 
+  // until sbt/sbt#2402 is fixed (https://github.com/sbt/sbt/issues/2402)
+  lazy val cleanBridge = TaskKey[Unit]("clean-sbt-bridge", "delete dotty-sbt-bridge cache")
```
Comment (MasseGuillaume) [inline project/Build.scala]: the convention is to use `a-b-c`
http://www.scala-sbt.org/0.13/sxr/sbt/Keys.scala.html#sbt.Keys.logLevel

Comment (MasseGuillaume) [inline project/Build.scala]: ah more or less: http://www.scala-sbt.org/0.13/sxr/sbt/Keys.scala.html#sbt.Keys.compileAnalysisFilename

:P. I will do as you said

File: project/Build.scala (hunk 8 above)
Comment (MasseGuillaume) [inline project/Build.scala]: done

File: project/Build.scala (hunk 11 above)
Comment (MasseGuillaume) [inline project/Build.scala]: done

**Comment by MasseGuillaume** [comment]
@smarter all issues resolved!

File: project/Build.scala (hunk 11 above)
Comment (smarter) [inline project/Build.scala]: Sorry if I wasn't totally clear, I meant that the `val` should be called `cleanSbtBridge`, but I'm fine with changing the key name too :)

**Comment by smarter** [comment]
Though please squash the last commit with the review fixes in a previous one if possible

**Comment by MasseGuillaume** [comment]
@smarter rebased

**Pull request merged**